OpenSim/Simulation/Test/testStatesTrajectory.cpp

```diff
@@ -508,6 +508,386 @@ void testCopying() {
+    // Using a custom tolerance.
+    SimTK_TEST(states.findIndexNearestBefore(0.499, 0.001) == 0);
+    SimTK_TEST(states.findIndexNearestBefore(0.495, 0.005) == 0);
+    SimTK_TEST(states.findIndexNearestBefore(1.2991, 0.001) == 1);
+    SimTK_TEST(states.findIndexNearestBefore(1.298999, 0.001) == 0);
+    SimTK_TEST(states.findIndexNearestBefore(4.099, 0.001) == 3);
+    SimTK_TEST(states.findIndexNearestBefore(4.098, 0.001) == 2);
+
+    SimTK_TEST(states.findIndexNearestAfter(0.501, 0.001) == 0);
+    SimTK_TEST(states.findIndexNearestAfter(0.503, 0.003) == 0);
+    // Past the tolerance, we get the next state.
+    SimTK_TEST(states.findIndexNearestAfter(0.50301, 0.003) == 1);
+    SimTK_TEST(states.findIndexNearestAfter(1.301, 0.001) == 1);
+    SimTK_TEST(states.findIndexNearestAfter(1.301001, 0.001) == 2);
+    SimTK_TEST(states.findIndexNearestAfter(4.1005, 0.001) == 3);
+    SimTK_TEST(states.findIndexNearestAfter(4.101, 0.001) == 3);
+
+    // Test exceptions with custom tolerance.
+    SimTK_TEST_MUST_THROW_EXC(states.findIndexNearestBefore(0.498, 0.001),
+            StatesTrajectory::TimeOutOfRange);
+    SimTK_TEST_MUST_THROW_EXC(states.findIndexNearestBefore(0.498999, 0.001),
+            StatesTrajectory::TimeOutOfRange);
+    SimTK_TEST_MUST_THROW_EXC(states.findIndexNearestAfter(4.102, 0.001),
+            StatesTrajectory::TimeOutOfRange);
+    SimTK_TEST_MUST_THROW_EXC(states.findIndexNearestAfter(4.10100001, 0.001),
+            StatesTrajectory::TimeOutOfRange);
+
+    // Make sure that we get the right state when there are multiple with
+    // the same time.
+    state.setTime(4.1); // Same time as before.
+    states.append(state); // 4
+    states.append(state); // 5
+    state.setTime(5.8);
+    states.append(state); // 6
+    states.append(state); // 7
+    SimTK_TEST(states.findIndexNearestAfter(4.0) == 3);
+    SimTK_TEST(states.findIndexNearestAfter(4.1) == 3);
+    SimTK_TEST(states.findIndexNearestBefore(5.0) == 5);
+    SimTK_TEST(states.findIndexNearestBefore(4.1) == 5);
+
+    SimTK_TEST(states.findIndexNearestAfter(4.1004999, 0.0005) == 3);
+    SimTK_TEST(states.findIndexNearestBefore(4.0995, 0.0005) == 5);
+
+    SimTK_TEST(states.findIndexNearestAfter(5.8) == 6);
+    SimTK_TEST(states.findIndexNearestBefore(5.8) == 7);
+    SimTK_TEST_MUST_THROW_EXC(states.findIndexNearestAfter(5.800001),
+            StatesTrajectory::TimeOutOfRange);
+
+    // Make sure we get the right state when there are multiple within
+    // the tolerance.
+    {
+        // Assume the user views the times with only two decimal places (say
+        // in a console or storage file); they'll see:
+        //      0.50, 1.30, 3.50, 3.50, 3.50, 4.10.
+        // Therefore, the appropriate tolerance
+        // is 0.01. If they ask for the nearest state before 3.50, they should
+        // get the state that is actually at 3.502, since, from the
+        // console/file, this appears to be the first state "before" 3.50.
```

Comment:
Based on our discussion yesterday, I'll change the behavior so you get 3.5 instead.